pkg/liquidity-mining/contracts/veBALDeploymentCoordinator.sol

```diff
@@ -0,0 +1,242 @@
+        return _balancerMinter;
+    }
+
+    function getCurrentDeploymentStage() external view returns (DeploymentStage) {
+        return _currentDeploymentStage;
+    }
+
+    function getActivationScheduledTime() external view returns (uint256) {
+        return _activationScheduledTime;
+    }
+
+    function getGaugeTypes() external view returns (GaugeTypes[] memory) {
+        return _gaugeTypes;
+    }
+
+    function performFirstStage() external nonReentrant {
+        // Check internal state
+        require(block.timestamp >= _activationScheduledTime, "Not ready for activation");
+        require(_currentDeploymentStage == DeploymentStage.PENDING, "First steap already performed");
+
+        // Check external state: we need admin permission on both the BAL token and the Authorizer
+        IAuthorizer authorizer = _vault.getAuthorizer();
+        require(_balancerToken.hasRole(_balancerToken.DEFAULT_ADMIN_ROLE(), address(this)), "Not BAL admin");
+        require(authorizer.canPerform(bytes32(0), address(this), address(0)), "Not Authorizer admin");
+
+        // Sanity checks
+        IGaugeController gaugeController = _balancerMinter.getGaugeController();
+        require(gaugeController.n_gauge_types() == 0, "Gauge types already set");
+
+        // Step 1: trigger BAL token admin migration, locking the BAL emissions forever.
+        //
+        // The BalancerTokenAdmin contracts needs admin permission over BAL in order to complete this process, and we
+        // need to be authorized to make the call.
+        _balancerToken.grantRole(_balancerToken.DEFAULT_ADMIN_ROLE(), _balancerTokenAdmin);
+        authorizer.grantRole(_balancerTokenAdmin.getActionId(IBalancerTokenAdmin.activate.selector), address(this));
+        _balancerTokenAdmin.activate();
+        // Balancer Token Admin activation automatically removes authority over the BAL token from all other accounts,
+        // so we don't need to renounce this permission.
+
+        // Step 2: grant BalancerMinter authority to mint BAL, as part of the Liquidity Mining program.
+        authorizer.grantRole(_balancerTokenAdmin.getActionId(IBalancerTokenAdmin.mint.selector), _balancerMinter);
+
+        // Step 3: setup the Liquidity Mining program by creating the different gauge types on the Gauge Controller.
+        //
+        // All gauge types are initially created with a weight of 0, which will allow for gauges to be deployed and LPs
+        // to vote and stake for them, without yet minting any BAL. This is intended to provide a grace period for LPs
+        // to migrate to the new system before it is fully activated. The gauge type weights will be set to their actual
+        // values on the last stage of deployment.
+        {
+            // Admin functions on the Gauge Controller have to be called via the the AuthorizerAdaptor, which acts as
+            // its admin.
+            IAuthorizerAdaptor authorizerAdaptor = gaugeController.admin();
```

Comment:
This is currently assuming that ownership over the `GaugeController` has been properly transferred to the `AuthorizerAdaptor`. The Curve contracts will initially have the deployer key set as the admin and there's a two step process to transfer ownership.

We should either have the coordinator make the AuthorizerAdaptor claim ownership over the various contracts in `performFirstStage` or add a `prepareForFirstStage` function which will just perform all of these ownership transfers to get the system in a "proper uninitialised state". I'd lean towards the second as we can do additional verification before hitting the button but it's not a huge deal.

Reply:
From the looks of things the only contracts for which we need to do this is `VotingEscrow` and `GaugeController`

Reply:
Addressed in #1171